File: backend/models/scan.py
from pydantic import BaseModel, HttpUrl
from typing import List, Dict, Optional

class ScanRequest(BaseModel):
    target_url: HttpUrl

class ReconData(BaseModel):
    tech_stack: List[str]
    subdomains: List[str]

class Vulnerability(BaseModel):
    name: str
    severity: str
    fix_snippet: str

class ScanResponse(BaseModel):
    target_url: str
    score: str               # "A", "B", "C", "D", "F"
    score_percentage: int    # 0 to 100
    vulnerabilities: List[Vulnerability]
    ai_remediation: str = None # AI-Powered Remediation
    mermaid_syntax: str      # Dynamic attack topology
    graph_data: dict = None  # Segment 3: React Flow Graph Object
    recon_data: Optional[ReconData] = None # Segment 5: Target Intelligence
    status: str
    message: str

File: frontend/src/types/scan.ts
import type { Edge, Node } from "@xyflow/react";

export interface GraphNodeData {
  label: string;
  compromised?: boolean;
  node_type?: string;
  defended?: boolean;
  threat_level?: string;
  [key: string]: unknown;
}

export interface GraphData {
  nodes: Array<Node<GraphNodeData>>;
  edges: Array<Edge>;
}

export interface ReconData {
  tech_stack: string[];
  subdomains: string[];
}

export interface ScanResult {
  id: string;
  target_url: string;
  score: string;
  score_percentage: number;
  vulnerabilities: Array<{ name: string; severity: string; fix_snippet: string }>;
  recon_data?: ReconData;
  graph_data: GraphData;
  mermaid_syntax: string;
  ai_remediation?: string;
}
